src/ripple/app/tests/Flow_test.cpp

```diff
@@ -1271,6 +1271,39 @@ struct Flow_test : public beast::unit_test::suite
             BEAST_EXPECT(offer[sfTakerPays] == USD (495));
         }
     }
+    void testSelfFundedXRPEndpoint (bool consumeOffer)
+    {
+        // Test that the deferred credit table is not bypassed for
+        // XRPEndpointSteps. If the account in the first step is sending XRP and
+        // that account also owns an offer that receives XRP, it should not be
+        // possible for that step to use the XRP received in the offer as part
+        // of the payment.
+        testcase("Self funded XRPEndpoint");
+
+        using namespace jtx;
+
+        Env env(*this, features(featureFlowV2));
+
+        // Need new behavior from `accountHolds`
+        auto const closeTime = amendmentRIPD1141SoTime() +
+            env.closed()->info().closeTimeResolution;
+        env.close(closeTime);
+
+        auto const alice = Account("alice");
+        auto const gw = Account("gw");
+        auto const USD = gw["USD"];
+
+        env.fund(XRP(10000), alice, gw);
+        env(trust(alice, USD(20)));
+        env(pay(gw, alice, USD(10)));
+        env(offer(alice, XRP(50000), USD(10)));
+
+        // Consuming the offer changes the owner count, which could also cause
+        // liquidity to decrease in the forward pass
+        auto const toSend = consumeOffer ? USD(10) : USD(9);
+        env(pay(alice, alice, toSend), path(~USD), sendmax(XRP(20000)),
+            txflags(tfPartialPayment | tfNoRippleDirect));
```

Comment:
@scottschurr `consumeOffer` adds to the test coverage because if the offer is consumed in the reverse pass, the owner count is decreased. When the first step is the limiting step it could compute the reserve with a decreased owner count, then when re-computing compute the reserve with the increased owner count.

The test is that the payment succeeds. The balances won't change. In debug mode an assert would fire, but even in release the payment would fail (without these fixes). Checking owner counts isn't helpful because the test is for owner count changes during the payment, not before or after.